ampd/src/sui/verifier.rs

```diff
@@ -0,0 +1,194 @@
+            return false;
+        }
+
+        serde_json::from_value::<ContractCall>(self.parsed_json.clone()).unwrap();
+
+        match serde_json::from_value::<ContractCall>(self.parsed_json.clone()) {
+            Ok(contract_call) => {
+                contract_call.source_id == msg.source_address
+                    && contract_call.destination_chain == msg.destination_chain.to_string()
+                    && contract_call.destination_address == msg.destination_address
+                    && contract_call.payload_hash == msg.payload_hash
+            }
+            _ => false,
+        }
+    }
+}
+
+fn get_event<'a>(
+    gateway_address: &SuiAddress,
+    transaction_block: &'a SuiTransactionBlockResponse,
+    event_seq: u64,
+) -> Option<&'a SuiEvent> {
+    if let Some(events) = transaction_block.events.as_ref() {
+        events.data.iter().find(|event| {
+            event.id.event_seq == event_seq
+                && <ObjectID as Into<SuiAddress>>::into(event.package_id) == *gateway_address
+        })
+    } else {
+        None
+    }
+}
+
+#[allow(dead_code)]
+pub fn verify_message(
+    gateway_address: &SuiAddress,
+    transaction_block: &SuiTransactionBlockResponse,
+    message: &Message,
+) -> bool {
+    match get_event(gateway_address, transaction_block, message.event_index) {
+        Some(event) => transaction_block.digest == message.tx_id && event == message,
+        None => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use ethers::abi::AbiEncode;
+    use move_core_types::language_storage::StructTag;
+    use sui_json_rpc_types::{SuiEvent, SuiTransactionBlockEvents, SuiTransactionBlockResponse};
+    use sui_types::{
+        base_types::{SuiAddress, TransactionDigest},
+        event::EventID,
+    };
+
+    use crate::handlers::sui_verify_msg::Message;
+    use crate::sui::verifier::verify_message;
+    use crate::types::{EVMAddress, Hash};
+
+    fn get_matching_msg_and_tx_block() -> (SuiAddress, SuiTransactionBlockResponse, Message) {
+        let gateway_address = SuiAddress::random_for_testing_only();
+
+        let msg = Message {
+            tx_id: TransactionDigest::random(),
+            event_index: 999,
+            source_address: SuiAddress::random_for_testing_only(),
+            destination_chain: "ethereum".parse().unwrap(),
+            destination_address: format!("0x{:x}", EVMAddress::random()).parse().unwrap(),
+            payload_hash: Hash::random(),
+        };
+
+        let json_str = format!(
+            r#"{{"destination_address": "{}", "destination_chain": "{}",  "payload": "[1,2,3]",
+            "payload_hash": "{}",  "source_id": "{}"}}"#,
+            msg.destination_address,
+            msg.destination_chain,
+            msg.payload_hash.encode_hex(),
+            msg.source_address
+        );
+        let parsed: serde_json::Value = serde_json::from_str(json_str.as_str()).unwrap();
```

Comment:
what does this represent?

Reply:
is the emitted CallContract event in serialized json value